Enable pagination of stats model queries and filtering/ordering by timestamp

nateyazdani opened the pull request:
Enable (optional) pagination of JSON-P stats model requests, using an API similar to datagrepper's. Also adds support to constrain stats model requests within a given time window and to order the results by time, either ascending or descending. Some open questions that we may want to consider:
1. Should and how might CSV results be paginated? The format does not lend itself as well to this sort of thing...
2. Should pagination be mandatory? If not, should the same format be enforced, so that all results are returned as a single page?

Commit:
Add SQLAlchemy query paginator class

Pull in a class to perform pagination on SQLAlchemy queries from the autocloud
project.

Commit:
Reword comment to fit in an 80-character line

Commit:
Reword URL endpoint docstrings

Reword the Flask route endpoint docstrings to reflect the availability of
content-types other than JSON-P.

Commit:
Allow basic query filtering from web interface

Use query string to allow basic filtering and ordering of the model query in the
web interface.

Commit:
Customize pagination

Refactors the class statscache.pagination.RangeBasedPagination into the simple
utility function statscache.utils.paginate() that has the minimum necessary code
for the behavior of the statscache API.

Commit:
Allow pagination of JSON-P model queries

Use the new statscache.utils.paginate() function to enable optional pagination
of stats model queries in JSON[-P].

Commit:
Correct behavior of 'paginate' URL argument

Actually check whether the 'paginate' query string argument equals 'true' or
'yes' in the stats model serialization endpoint before deciding to enable
pagination.

ralphbean commented on statscache/app.py:
```diff
@@ -52,18 +57,85 @@ def plugin_index():
+                page_len,
+                '?'.join([flask.request.base_url, query_string])
+            )
+            data = ("{ 'items': %s,"
```
While this looks like it will function correctly, it feels kind of odd to not be using `json.dumps` here.

nateyazdani replied:
Yes, I agree, but to use `json.dumps` would require unnecessarily parsing the JSON string returned by `plugin.to_json` only to have `json.dumps` re-compute the same JSON string. When I looked into it, I couldn't find any way to make `json.dumps` treat a string as pre-computed JSON.

ralphbean replied:
Good to know the reasoning.  Can you add this as an inline comment in the source?

ralphbean commented:
Sorry this sat idle in the queue for so long.  It looks good to me except that those formatting lines give me pause.

ralphbean commented on statscache/app.py:
```diff
@@ -52,18 +57,85 @@ def plugin_index():
+                int(flask.request.args.get('rows_per_page',
+                                           default=DEFAULT_ROWS_PER_PAGE))
+            )
+            query_string = '&'.join([
```
It might be nice to use `urllib.urlencode` here:

``` python
>>> urllib.urlencode({'foo': 'bar', 'baz': 'wat'})
'foo=bar&baz=wat'
```

nateyazdani replied:
That is a good idea, thanks

Commit:
Use urllib for query string formatting

Rather than raw string manipulation, use urllib.urlencode to (safely) construct
URL query strings.

nateyazdani commented:
No worries! You do raise some good points, though. If I have addressed them well enough, then I could also use your input on my two points in the PR description (whether to paginate CSV and whether to use pagination format for all JSON requests).

ralphbean commented:
Oh, I missed these.  I saw the PR had been idle for a few days and I hastily jumped to the code.

> Should and how might CSV results be paginated? The format does not lend itself as well to this sort of thing...

I'd say yes, if only because it makes the API symmetric.

> Should pagination be mandatory? If not, should the same format be enforced, so that all results are returned as a single page?

Let's say _no_ for now and see how it goes.  If we find we're being DoS'd into the ground then we can make it mandatory.  Hopefully statscache will be the fast complement to datagrepper's slowe, paging-is-mandatory service.

nateyazdani commented:
No sweat, appreciate the enthusiasm :)

Sounds good on both points. Do you have any ideas or suggestions for how to paginate CSV?

ralphbean commented:
Guh, there's no standard for "comments" in a csv file where we could stuff some metadata about how many rows-per-page and which page the document represents, etc..

How about putting that data in HTTP response headers?  `X-statscache-page: 22`

_If_ we do that, we should copy it for the JSON renderer so that, even while JSON has that paging information included in the payload body it also includes it as response headers (for symmetry between the formats).

Commit:
Use HTTP headers to control pagination

Rather than supplementing JSON (and only JSON) responses with pagination-related
metadate, pass this information along through response headers. This also makes
it possible to support CSV pagination in a consistent manner, which has been
done. In order to allow web clients to access this metadata, the HTTP headers
"Access-Control-Allow-Origin" and "Access-Control-Expose-Headers" are set
appropriately for paginated responses.

Commit:
Correct outdated comment

nateyazdani commented:
Updated the code to implement pagination metadata through HTTP headers. Currently, I'm using `Link` to give the next and previous URLs (which is fairly standard, I have learned), `X-Link-Number` for the current page number, and `X-Link-Count` to give the total number of pages. Also turned on pagination for CSV and enabled the `Access-Control-*` headers so that this will actually be usable.

Commit:
Correct use of mutable default argument

Commit:
Correct typo in last commit

ralphbean commented:
Looks good!

Merge it up -- :bomb:s away!
